[embedded] Fix associated type conformances for specialized witness tables

aschwaighofer opened the pull request.

eeckstein committed:
embedded: don't re-abstract witness method calls for existentials

Don't convert indirect to direct arguments. This is needed to support general existentials in embedded swift.

eeckstein committed:
SIL: add some APIs for `InitExistentialAddrInst`

* `var conformances: ConformanceArray`
* `var formalConcreteType: CanonicalType`

eeckstein committed:
MandatoryPerformanceOptimizations: specialize witness tables for general existentials

aschwaighofer committed:
[embedded] A few more test cases for existential casting

aschwaighofer committed:
[embedded] Tighten the check which existentials we support in embedded with existentials

aschwaighofer committed:
[embedded] Allow casting to tuples

aschwaighofer committed:
[embedded] One more test

eeckstein committed:
AST: add `var ProtocolDecl.isMarkerProtocol`

aschwaighofer committed:
[embedded] Fix associate type conformances involving specialized conformances

aschwaighofer committed:
[embedded] Existentials: enable tests that now pass

aschwaighofer commented:
@swift-ci test

aschwaighofer committed:
[embedded] Enable all tests in general-existentials1.swift

aschwaighofer commented:
@swift-ci test

slavapestov commented on lib/IRGen/GenProto.cpp:
```diff
@@ -1751,6 +1751,10 @@ class AccessorConformanceInfo : public ConformanceInfo {
       auto typeWitness = Conformance.getTypeWitness(assocType);
+      if (langOpts.hasFeature(Feature::EmbeddedExistentials) &&
+          SILWT->isSpecialized()) {
+        typeWitness = entry.getAssociatedTypeWitness().Witness;
```
This doesn't make sense, why is Conformance.getTypeWitness() returning the wrong thing? Is Conformance is not a SpecializedProtocolConformance? If not, it should be fixed to match up and the conditional removed.

aschwaighofer committed:
[embedded] Add support for storing/casting function types

aschwaighofer commented:
Failed Tests (1):
14:44:53    Swift(iphonesimulator-x86_64) :: TBD/linker-directives-ld-previous-ios.swift

aschwaighofer committed:
[embedded] Use the type witness from the specialized conformance

rather then the specialized conformance's root conformance (which is the  variable's value)
